### ISSUE TITLE
AWS docker registry support

### ISSUE THREAD

**Issue by petrovicboban** [HubSpot/Singularity#1710]
It seems that pulling docker images from AWS private docker registries is not supported.    
This is the part of executor.java.log:
```
com.hubspot.singularity.runner.base.shared.ProcessFailedException: Could not pull docker image
	at com.hubspot.singularity.executor.task.SingularityExecutorTaskProcessBuilder.call(SingularityExecutorTaskProcessBuilder.java:68) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.task.SingularityExecutorTaskProcessBuilder.call(SingularityExecutorTaskProcessBuilder.java:24) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: com.spotify.docker.client.DockerException: java.util.concurrent.ExecutionException: com.spotify.docker.client.DockerRequestException: Request error: POST unix://localhost:80/images/create?fromImage=_accountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_&tag=latest: 500
	at com.hubspot.singularity.executor.utils.DockerUtils.pull(DockerUtils.java:64) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.task.SingularityExecutorTaskProcessBuilder.call(SingularityExecutorTaskProcessBuilder.java:66) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	... 5 common frames omitted
Caused by: java.util.concurrent.ExecutionException: com.spotify.docker.client.DockerRequestException: Request error: POST unix://localhost:80/images/create?fromImage=_accountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_&tag=latest: 500
	at com.github.rholder.retry.Retryer$ExceptionAttempt.<init>(Retryer.java:206) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.github.rholder.retry.Retryer.call(Retryer.java:135) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.utils.DockerUtils.callWithRetriesAndTimeout(DockerUtils.java:122) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.utils.DockerUtils.pull(DockerUtils.java:62) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	... 6 common frames omitted
Caused by: com.spotify.docker.client.DockerRequestException: Request error: POST unix://localhost:80/images/create?fromImage=_accountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_&tag=latest: 500
	at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1133) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1082) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.spotify.docker.client.DefaultDockerClient.pull(DefaultDockerClient.java:665) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.spotify.docker.client.DefaultDockerClient.pull(DefaultDockerClient.java:650) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.utils.DockerUtils$2.call(DockerUtils.java:56) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.hubspot.singularity.executor.utils.DockerUtils$2.call(DockerUtils.java:54) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
	at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:1.8.0_131]
	at java.util.concurrent.FutureTask.get(FutureTask.java:206) ~[na:1.8.0_131]
	at com.google.common.util.concurrent.SimpleTimeLimiter.callWithTimeout(SimpleTimeLimiter.java:130) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.github.rholder.retry.AttemptTimeLimiters$FixedAttemptTimeLimit.call(AttemptTimeLimiters.java:107) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at com.github.rholder.retry.Retryer.call(Retryer.java:132) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	... 8 common frames omitted
Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:968) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:795) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.JerseyInvocation.access$500(JerseyInvocation.java:91) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.JerseyInvocation$5.completed(JerseyInvocation.java:756) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:189) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:74) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.ClientRuntime$1.run(ClientRuntime.java:171) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:320) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:201) ~[SingularityExecutor-0.13.0-shaded.jar:0.13.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_131]
	... 4 common frames omitted
```

I'm using singulariy version 0.13.  
Is this solved in newer versions, or is there something I'm missing?

**Comment by ssalinas** [comment]
I wouldn't expect pulling from an aws registry to be any different than pulling from a regular one. This was just trying to pull `_accountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_:latest` correct? Did you `docker login` on the host first so that it was authenticated with the registry?

**Comment by petrovicboban** [comment]
Yes, I did. After authenticating from cli with `docker login`, using local root account, I can pull images from aws. But it seems that singularity executor cannot, and I'm pretty sure that it runs with root account. Where should I look now?

**Comment by petrovicboban** [comment]
It's interesting that the same error shows up even if I pull image manually, and not set force image pull:     
```
container {
  type: DOCKER
  docker {
    image: "_acountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_:latest"
    network: NONE
    privileged: false
    force_pull_image: false
  }
}
```

**Comment by petrovicboban** [comment]
Is there more detailed logs for spotify docker client?  
  
```
Caused by: com.spotify.docker.client.DockerRequestException: Request error: POST unix://localhost:80/images/create?fromImage=_acountID_.dkr.ecr.eu-west-2.amazonaws.com/_test_&tag=latest: 500
...omited
Caused by: javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
```

**Comment by petrovicboban** [comment]
When using mesos executor, archiving `.docker`  with `config.json` in it, and using that archive as artifact, does the job.  But making singularity executor to work with `.docker/config.json` looks almost imposible.

**Comment by ssalinas** [comment]
We have credential settings available in the executor config yaml https://github.com/HubSpot/Singularity/blob/master/SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorConfiguration.java#L210 -> https://github.com/HubSpot/Singularity/blob/master/SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorDockerAuthConfig.java

**Comment by petrovicboban** [comment]
Thanks a lot, it works! Although, I need to update executor configuration each time login expires. Not ideal, but it can be handled.

**Comment by petrovicboban** [comment]
if anyone needs, I wrote simple script which does job for me:  
```
#!/bin/bash

/usr/local/bin/aws ecr get-login --no-include-email --region eu-west-2 | bash
line=$(cat .docker/config.json | grep auth | awk '{print $2}' | sed 's/"//g' | sed 's/{//g' | base64 --decode | awk -F: '{print "  password: " $2}')
sed -i "s/  password.*/$line/g" /etc/singularity.executor.yaml
```

Putting this in cron to be executed every couple of hours is enough.

**Comment by petrovicboban** [comment]
This seems to be fixed in #1534 .
Closing.

**Comment by serverhorror** [comment]
Hi,
Sorry to revive this. Is it supposed to also work with awslabs/amazon-ecr-credential-helper?
